### PR TITLE
CMake: Support dynamic codegen outputs

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #68462
* #68693
* #68692
* #68691
* #68697
* #68690
* #68689
* #68714
* #68688
* #68687
* #68247
* __->__ #68246

Currently the codegen produces a list of output files at CMake
configuration time and the build system has no way of knowing if the
outputs change. So if that happens, you basically need to delete the
build folder and re-run from scratch.

Instead, this generates the output list every time the code generation
is run and changes the output to be a `.cmake` file that gets included
in the main cmake configuration step. That means the build system
knows to re-run cmake automatically if a new output is added. So, for
example you could change the number of shards that `Operators.cpp` is
split into and it all just works transparently to the user.

Differential Revision: [D32596268](https://our.internmc.facebook.com/intern/diff/D32596268)

### CODE REVIEW

File: .circleci/docker/build.sh [hunk 1]
```diff
@@ -165,7 +165,7 @@ case "$image" in
   pytorch-linux-xenial-py3-clang5-android-ndk-r19c)
     ANACONDA_PYTHON_VERSION=3.6
     CLANG_VERSION=5.0
-    CMAKE_VERSION=3.10.3
+    CMAKE_VERSION=3.13.5
```
Comment (peterbell10) [inline .circleci/docker/build.sh:168]: cc @malfet 

There is a bug in the Ninja generator where files considered inputs to the cmake command couldn't be generated by another build step. The fix requires 3.13 but 3.10.3 is still sufficient for other cmake generators e.g. makefiles
https://gitlab.kitware.com/cmake/cmake/-/issues/18584

**Review by albanD — COMMENTED** [review]
The main difference after this PR is that changes that used to require a clean of the whole build will now properly detect them and re-run the cmake?

**Comment by peterbell10** [comment]
Yes, that's the idea.

The codegen output lists are now updated every time codegen gets run, not just on first configuration. Further, the `build.ninja` file now lists the file lists as dependencies for the cmake configuration so it will get re-run when new outputs files are added.

**Review by albanD — COMMENTED** [review]
Looks good. Just the cmake version bump to confirm.

File: .circleci/docker/build.sh (hunk 1 above)
Comment (malfet) [reply .circleci/docker/build.sh:168]: Please make sure to update https://github.com/pytorch/pytorch/blob/39ec0f321b2f5bd218dd1136ee08a951d876a33f/CMakeLists.txt#L1
as well

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596268).

**Comment by peterbell10** [comment]
@albanD the cmake version bump has been merged in #68731 and I've rebased on top of it. Unfortunately some of the CI is broken on master at the moment for unrelated reasons.

**Comment by peterbell10** [comment]
Rebased on a newer commit and CI is green again.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596268).

**Comment by albanD** [comment]
@peterbell10 did you update the main CMake file as per Nikita's request? I don't think I see it here.

**Comment by peterbell10** [comment]
@albanD the cmake version bump was merged separately in #68731 and includes an error if cmake is too old.

**Comment by albanD** [comment]
Ho perfect! Thanks

**Review by albanD — APPROVED** [review]

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596268).

**Pull request closed without merging**